File: src/main/java/destiny/astrology/classical/rules/accidentalDignities/Oriental.java
/**
 * @author smallufo
 * Created on 2007/12/29 at 上午 5:10:06
 */
package destiny.astrology.classical.rules.accidentalDignities;

import destiny.astrology.*;
import org.jetbrains.annotations.NotNull;
import org.jooq.lambda.tuple.Tuple;
import org.jooq.lambda.tuple.Tuple2;

import java.util.Optional;

/** Mars, Jupiter, or Saturn oriental of (rising before) the Sun.
 * 火星、木星、土星 是否 東出 於 太陽
 * */
public final class Oriental extends Rule {

  public Oriental() {
  }

  @Override
  protected Optional<Tuple2<String, Object[]>> getResult(Planet planet, @NotNull Horoscope h) {

    return h.getPositionOpt(planet)
      .filter(pos -> planet == Planet.MARS || planet == Planet.JUPITER || planet == Planet.SATURN)
      .map(Position::getLng).flatMap(planetDegree ->
        h.getPositionOpt(Planet.SUN)
          .map(Position::getLng)
          .filter(sunDegree -> Horoscope.isOriental(planetDegree, sunDegree))
          .map(sunDegree -> Tuple.tuple("comment", new Object[]{planet}))
    );
  }

}

File: src/main/java/destiny/astrology/classical/rules/accidentalDignities/Oriental.kt
/**
 * @author smallufo
 * Created on 2007/12/29 at 上午 5:10:06
 */
package destiny.astrology.classical.rules.accidentalDignities

import destiny.astrology.Horoscope
import destiny.astrology.IHoro
import destiny.astrology.Planet
import destiny.astrology.Planet.*

/** Mars, Jupiter, or Saturn oriental of (rising before) the Sun.
 * 火星、木星、土星 是否 東出 於 太陽
 */
class Oriental : Rule() {

  override fun getResult(planet: Planet, h: IHoro): Pair<String, Array<Any>>? {
    val planetDegree: Double? = arrayOf(MARS, JUPITER, SATURN)
      .takeIf { it.contains(planet) }
      ?.let { h.getPosition(planet) }?.lng

    val sunDegree: Double? = h.getPosition(SUN)?.lng

    return if (sunDegree != null && planetDegree != null && IHoro.isOriental(planetDegree, sunDegree)) {
      "comment" to arrayOf<Any>(planet)
    } else {
      null
    }
  }
}
